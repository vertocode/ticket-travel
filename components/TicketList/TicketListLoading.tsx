import {ReactElement} from 'react'
import '@/styles/components/TicketList/TicketListLoading.scss'

const TicketListLoading = (): ReactElement => {
    const numberOfTicketsLoading = 6
    const loadingTickets = new Array(numberOfTicketsLoading).fill('')

    return (
        <ul className="ticket-list-loading">
            {loadingTickets.map((_, index) => (
                <li className="ticket-card-loading" key={`loading-ticket-card-${index}`}>
                    <div className="ticket-image-loading">
                        <span className="skeleton-box" style={{ width: '200px', height: '130px' }}></span>
                    </div>
                    <div className="ticket-loading-body">
                        <h3>
                            <span className="skeleton-box" style={{ width: '80%' }}></span>
                            <span className="skeleton-box" style={{ width: '55%' }}></span>
                        </h3>
                        <p>
                            <span className="skeleton-box" style={{ width: '30%' }}></span>
                        </p>
                    </div>
                </li>
            ))}
        </ul>
    )
}

export default TicketListLoading