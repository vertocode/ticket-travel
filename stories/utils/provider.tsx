import type { ReactElement } from 'react'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { Provider as ProviderStore } from 'react-redux'
import cartReducer from '@/lib/features/cart/cartSlice'

export function Provider(component: ReactElement) {
	const store = configureStore({
		reducer: combineReducers({ cartReducer })
	})

	return <ProviderStore store={store}>{component}</ProviderStore>
}