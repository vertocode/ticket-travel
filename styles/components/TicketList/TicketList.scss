.ticket-list {
  display: flex;
  flex-direction: column;
  gap: 1em;
}

.empty-list-message {
  font-size: 1.5em;
  text-align: center;
  background-color: #0A2156;
  color: #fff;
  padding: 1em;
  border-radius: 5px;
}