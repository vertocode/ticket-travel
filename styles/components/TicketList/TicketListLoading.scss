.ticket-list-loading {
  margin: 0;
  list-style: none;
  width: 100%;

  .ticket-card-loading {
    margin-bottom: 1em;
    display: grid;
    grid-template-columns: 1fr 3fr;
    padding: 2em;
    background-color: #fff;

    .ticket-image-loading {
      width: 30%;
      height: 100px;
      margin-right: 1em;
    }

    .skeleton-box {
      display: inline-block;
      height: 1em;
      width: 100%;
      position: relative;
      overflow: hidden;
      background-color: #DDDBDD;

      &::after {
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        transform: translateX(-100%);
        background-image: linear-gradient(
          90deg,
          rgba(#fff, 0) 0,
          rgba(#fff, 0.2) 20%,
          rgba(#fff, 0.5) 60%,
          rgba(#fff, 0)
        );
        animation: shimmer 3s infinite;
        content: '';
      }

      @keyframes shimmer {
        100% {
          transform: translateX(100%);
        }
      }
    }
  }
}


