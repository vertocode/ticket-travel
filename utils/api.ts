const API_URL = `https://65b98494b71048505a8aea91.mockapi.io/api/v1`

export const get = async (endpoint: string, options?: { headers?: any, revalidate?: number }) => {
  const response = await fetch(`${API_URL}${endpoint}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers
    },
    next: {
      revalidate: options?.revalidate
    }
  })
  return response.json()
}
