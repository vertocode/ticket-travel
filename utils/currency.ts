export const formatCurrency = (currency: number): string => {
    return currency.toLocaleString('pt-BR', { style: 'currency', currency: 'BRL' });
}