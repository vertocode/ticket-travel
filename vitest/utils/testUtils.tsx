import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'

import type { AppStore, RootState } from '@/lib/store'

import cartReducer from '@/lib/features/cart/cartSlice'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>
    store?: AppStore
}

export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {},
		// Automatically create a store instance if no store was passed in
		store = configureStore({ reducer: combineReducers({ cartReducer }), preloadedState }),
		...renderOptions
	}: ExtendedRenderOptions = {}
) {
	// there is a recommendation by testing-library docs to use userEvent.setup() before anything in the tests.
	// reference: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
	const user = userEvent.setup()

	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return <Provider store={store}>{children}</Provider>
	}

	// Return an object with the store and all of RTL's query functions
	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }), user }
}