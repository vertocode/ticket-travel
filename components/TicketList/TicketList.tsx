'use client'
import { ReactElement } from 'react'

// styles
import '@/styles/components/TicketList/TicketList.scss'

// hocks
import { useTicketsFetch } from '@/hooks/useTicketsFetch'

// components
import TicketListLoading from '@/components/TicketList/TicketListLoading'
import TicketCard from '@/components/TicketList/TicketCard/TicketCard'
import Pagination from '@/components/TicketList/Pagination/Pagination'
import { Suspense } from 'react'
import Skeleton from '@mui/material/Skeleton'

const TicketList = (): ReactElement => {
	const { tickets, loading, error, showPagination } = useTicketsFetch()

	return (
		<div className="ticket-list">
			{ !loading && error && <p className="error">{ error }</p> }
			{ loading && !error
				? <TicketListLoading />
				: tickets.map(ticket => <TicketCard key={ticket.id} ticket={ticket} />)
			}
			{ showPagination && tickets.length && (
				<Suspense fallback={ <Skeleton /> }>
					<Pagination />
				</Suspense>
			)}
			{ !loading && !tickets.length && !error && (
				<p className="empty-list-message">
					No momento, não há ingressos disponíveis. Experimente ajustar seus critérios de busca para encontrar novas oportunidades emocionantes!
				</p>
			)}
		</div>
	)
}

export default TicketList