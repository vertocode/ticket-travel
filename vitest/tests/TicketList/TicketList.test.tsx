import { expect, test, vi, beforeAll, describe } from 'vitest'
import { renderWithProviders } from '@/vitest/utils/testUtils'
import TicketList from '@/components/TicketList/TicketList'

const { container, getByTestId } =  renderWithProviders(<TicketList />)

describe.concurrent('TicketList', () => {
	beforeAll(() => {
		vi.mock('next/navigation', () => {
			const actual = vi.importActual('next/navigation')
			return {
				...actual,
				useRouter: vi.fn(() => ({
					push: vi.fn(),
				})),
				useSearchParams: vi.fn(() => ({
					get: vi.fn(),
				})),
				usePathname: vi.fn(),
			}
		})
	})

	test('should match snapshot when there are tickets', () => {
		expect(container).toMatchSnapshot()
	})

	test('should show the pagination when there are more than 6 tickets', async () => {
		// wait the loading finishes
		await new Promise(resolve => setTimeout(resolve, 2000))
		expect(await getByTestId('pagination')).toBeDefined()
	})
})
