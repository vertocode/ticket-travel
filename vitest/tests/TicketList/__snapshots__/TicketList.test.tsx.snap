// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TicketList > should show the pagination when there are more than 6 tickets 1`] = `
<div>
  <div
    class="ticket-list"
  >
    <ul
      class="ticket-list-loading"
    >
      <li
        class="ticket-card-loading"
      >
        <div
          class="ticket-image-loading"
        >
          <span
            class="skeleton-box"
            style="width: 200px; height: 130px;"
          />
        </div>
        <div
          class="ticket-loading-body"
        >
          <h3>
            <span
              class="skeleton-box"
              style="width: 80%;"
            />
            <span
              class="skeleton-box"
              style="width: 55%;"
            />
          </h3>
          <p>
            <span
              class="skeleton-box"
              style="width: 30%;"
            />
          </p>
        </div>
      </li>
      <li
        class="ticket-card-loading"
      >
        <div
          class="ticket-image-loading"
        >
          <span
            class="skeleton-box"
            style="width: 200px; height: 130px;"
          />
        </div>
        <div
          class="ticket-loading-body"
        >
          <h3>
            <span
              class="skeleton-box"
              style="width: 80%;"
            />
            <span
              class="skeleton-box"
              style="width: 55%;"
            />
          </h3>
          <p>
            <span
              class="skeleton-box"
              style="width: 30%;"
            />
          </p>
        </div>
      </li>
      <li
        class="ticket-card-loading"
      >
        <div
          class="ticket-image-loading"
        >
          <span
            class="skeleton-box"
            style="width: 200px; height: 130px;"
          />
        </div>
        <div
          class="ticket-loading-body"
        >
          <h3>
            <span
              class="skeleton-box"
              style="width: 80%;"
            />
            <span
              class="skeleton-box"
              style="width: 55%;"
            />
          </h3>
          <p>
            <span
              class="skeleton-box"
              style="width: 30%;"
            />
          </p>
        </div>
      </li>
      <li
        class="ticket-card-loading"
      >
        <div
          class="ticket-image-loading"
        >
          <span
            class="skeleton-box"
            style="width: 200px; height: 130px;"
          />
        </div>
        <div
          class="ticket-loading-body"
        >
          <h3>
            <span
              class="skeleton-box"
              style="width: 80%;"
            />
            <span
              class="skeleton-box"
              style="width: 55%;"
            />
          </h3>
          <p>
            <span
              class="skeleton-box"
              style="width: 30%;"
            />
          </p>
        </div>
      </li>
      <li
        class="ticket-card-loading"
      >
        <div
          class="ticket-image-loading"
        >
          <span
            class="skeleton-box"
            style="width: 200px; height: 130px;"
          />
        </div>
        <div
          class="ticket-loading-body"
        >
          <h3>
            <span
              class="skeleton-box"
              style="width: 80%;"
            />
            <span
              class="skeleton-box"
              style="width: 55%;"
            />
          </h3>
          <p>
            <span
              class="skeleton-box"
              style="width: 30%;"
            />
          </p>
        </div>
      </li>
      <li
        class="ticket-card-loading"
      >
        <div
          class="ticket-image-loading"
        >
          <span
            class="skeleton-box"
            style="width: 200px; height: 130px;"
          />
        </div>
        <div
          class="ticket-loading-body"
        >
          <h3>
            <span
              class="skeleton-box"
              style="width: 80%;"
            />
            <span
              class="skeleton-box"
              style="width: 55%;"
            />
          </h3>
          <p>
            <span
              class="skeleton-box"
              style="width: 30%;"
            />
          </p>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
